File: apps/eoc/src/lib/actions/micro-live.ts
"use server";

/**
 * Phase 13 — Micro-Live Institution server actions.
 *
 * These actions proxy to the deny-by-default micro-live API. None of them
 * can activate real live trading: the backend has no reachable code path
 * to MICRO_LIVE_ACTIVE, credential endpoints never accept or return
 * secret values (reference names only), and dry-run validation never
 * submits an order.
 */

import type { ActionResult } from "@/lib/actions/auth";
import { apiFetch } from "@/lib/server/api";
import type { DryRunOrderResult } from "@/lib/types";

export async function transitionActivationAction(
  formData: FormData,
): Promise<ActionResult> {
  const target_state = String(formData.get("target_state") ?? "").trim();
  const reason = String(formData.get("reason") ?? "").trim();
  if (!target_state || !reason) {
    return { ok: false, message: "Target state and reason are required." };
  }
  try {
    await apiFetch("/api/v1/micro-live/activation/transition", {
      method: "POST",
      body: { target_state, reason },
      idempotencyKey: crypto.randomUUID(),
    });
    return { ok: true, message: `Transition to ${target_state} submitted.` };
  } catch (err) {
    return {
      ok: false,
      message: err instanceof Error ? err.message : "Transition failed.",
    };
  }
}

export async function createCredentialReferenceAction(
  formData: FormData,
): Promise<ActionResult> {
  const provider_key = String(formData.get("provider_key") ?? "").trim();
  const ref_name = String(formData.get("ref_name") ?? "").trim();
  const purpose = String(formData.get("purpose") ?? "").trim();
  if (!provider_key || !ref_name || !purpose) {
    return {
      ok: false,
      message: "Provider, reference name, and purpose are required.",
    };
  }
  try {
    await apiFetch("/api/v1/micro-live/credential-references", {
      method: "POST",
      body: { provider_key, ref_name, purpose },
    });
    return {
      ok: true,
      message: `Credential reference ${ref_name} registered (presence-only; no value stored).`,
    };
  } catch (err) {
    return {
      ok: false,
      message: err instanceof Error ? err.message : "Create reference failed.",
    };
  }
}

export async function validateCredentialReferenceAction(
  formData: FormData,
): Promise<ActionResult> {
  const id = String(formData.get("reference_id") ?? "");
  if (!id) return { ok: false, message: "Reference id is required." };
  try {
    await apiFetch(`/api/v1/micro-live/credential-references/${id}/validate`, {
      method: "POST",
    });
    return { ok: true, message: "Presence check completed (no value exposed)." };
  } catch (err) {
    return {
      ok: false,
      message: err instanceof Error ? err.message : "Validation failed.",
    };
  }
}

export async function setKillSwitchAction(
  formData: FormData,
): Promise<ActionResult> {
  const scope_type = String(formData.get("scope_type") ?? "").trim();
  const scope_id = String(formData.get("scope_id") ?? "").trim() || null;
  const active = formData.get("active") === "on";
  const reason = String(formData.get("reason") ?? "").trim() || null;
  if (!scope_type) return { ok: false, message: "Scope type is required." };
  try {
    await apiFetch("/api/v1/micro-live/kill-switches", {
      method: "POST",
      body: { scope_type, scope_id, active, reason },
      idempotencyKey: crypto.randomUUID(),
    });
    return {
      ok: true,
      message: `Kill switch (${scope_type}) set to ${active ? "ACTIVE" : "cleared"}.`,
    };
  } catch (err) {
    return {
      ok: false,
      message: err instanceof Error ? err.message : "Kill switch update failed.",
    };
  }
}

export async function putCapitalPolicyAction(
  formData: FormData,
): Promise<ActionResult> {
  const fields = [
    "max_deployable_capital",
    "max_order_notional",
    "max_daily_loss",
    "max_concurrent_exposure",
    "max_provider_exposure",
    "max_strategy_exposure",
  ] as const;
  const body: Record<string, string> = {};
  for (const f of fields) {
    const v = String(formData.get(f) ?? "").trim();
    if (!v) return { ok: false, message: `${f} is required.` };
    body[f] = v;
  }
  try {
    await apiFetch("/api/v1/micro-live/capital-policy", {
      method: "PUT",
      body,
    });
    return { ok: true, message: "Micro-capital policy updated." };
  } catch (err) {
    return {
      ok: false,
      message: err instanceof Error ? err.message : "Policy update failed.",
    };
  }
}

export async function createReconciliationRunAction(
  formData: FormData,
): Promise<ActionResult> {
  const provider_key = String(formData.get("provider_key") ?? "").trim();
  const authoritative_raw = String(formData.get("authoritative_state") ?? "");
  const comparison_raw = String(formData.get("comparison_state") ?? "");
  if (!provider_key || !authoritative_raw || !comparison_raw) {
    return {
      ok: false,
      message: "Provider and both fixture states are required.",
    };
  }
  let authoritative_state: unknown;
  let comparison_state: unknown;
  try {
    authoritative_state = JSON.parse(authoritative_raw);
    comparison_state = JSON.parse(comparison_raw);
  } catch {
    return { ok: false, message: "Fixture states must be valid JSON." };
  }
  try {
    await apiFetch("/api/v1/micro-live/reconciliation/runs", {
      method: "POST",
      body: { provider_key, authoritative_state, comparison_state },
    });
    return { ok: true, message: "Reconciliation run completed." };
  } catch (err) {
    return {
      ok: false,
      message: err instanceof Error ? err.message : "Reconciliation run failed.",
    };
  }
}

export async function dryRunValidateOrderAction(
  formData: FormData,
): Promise<ActionResult & { result?: DryRunOrderResult }> {
  const quantity = String(formData.get("quantity") ?? "").trim();
  const reference_price = String(formData.get("reference_price") ?? "").trim();
  if (!quantity || !reference_price) {
    return { ok: false, message: "Quantity and reference price are required." };
  }
  try {
    const result = await apiFetch<DryRunOrderResult>(
      "/api/v1/micro-live/dry-run/validate-order",
      { method: "POST", body: { quantity, reference_price } },
    );
    return {
      ok: true,
      message: result.would_be_allowed
        ? "Dry-run: would be allowed under current policy (no order submitted)."
        : `Dry-run: would be BLOCKED — ${result.blocking_codes.join(", ")}`,
      result,
    };
  } catch (err) {
    return {
      ok: false,
      message: err instanceof Error ? err.message : "Dry-run validation failed.",
    };
  }
}

File: apps/api/app/api/v1/micro_live.py
"""Micro-Live Institution HTTP API (Phase 13) — deny-by-default.

No endpoint in this router can activate live trading, return a credential
value, or cause a real order to be submitted.
"""

from __future__ import annotations

import uuid

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from app.api.deps import RequireAnyAuthenticatedRead, RequireFounder, RequireFounderOrOperator
from app.db.session import get_db
from app.schemas.micro_live import (
    ActivationStateRead,
    ActivationTransitionRead,
    ActivationTransitionRequest,
    AdapterRead,
    CredentialReferenceCreate,
    CredentialReferenceRead,
    DryRunOrderResult,
    DryRunOrderValidate,
    KillSwitchRead,
    KillSwitchSet,
    MicroCapitalPolicyRead,
    MicroCapitalPolicyUpsert,
    MicroLiveStatusRead,
    ReconciliationDiscrepancyRead,
    ReconciliationRunCreate,
    ReconciliationRunRead,
)
from app.services.auth_service import AuthenticatedPrincipal, AuthError
from app.services.kill_switch_service import KillSwitchError, KillSwitchService
from app.services.live_activation_service import LiveActivationError, LiveActivationService
from app.services.micro_capital_service import MicroCapitalError, MicroCapitalService
from app.services.micro_live_service import MicroLiveService
from app.services.reconciliation_service import ReconciliationError, ReconciliationService

router = APIRouter(prefix="/api/v1/micro-live", tags=["micro-live"])

_ERROR_STATUS = {
    "invalid_state": status.HTTP_400_BAD_REQUEST,
    "invalid_transition": status.HTTP_400_BAD_REQUEST,
    "founder_required": status.HTTP_403_FORBIDDEN,
    "credentials_required": status.HTTP_400_BAD_REQUEST,
    "reason_required": status.HTTP_400_BAD_REQUEST,
    "live_execution_not_certified": status.HTTP_403_FORBIDDEN,
    "corrupt_state": status.HTTP_500_INTERNAL_SERVER_ERROR,
    "invalid_reference": status.HTTP_400_BAD_REQUEST,
    "reference_not_found": status.HTTP_404_NOT_FOUND,
    "invalid_scope": status.HTTP_400_BAD_REQUEST,
    "not_found": status.HTTP_404_NOT_FOUND,
    "policy_missing": status.HTTP_400_BAD_REQUEST,
    "invalid_policy": status.HTTP_400_BAD_REQUEST,
    "run_not_found": status.HTTP_404_NOT_FOUND,
}


def _http(exc: Exception) -> HTTPException:
    if isinstance(exc, AuthError):
        return HTTPException(status_code=403, detail=str(exc))
    if isinstance(
        exc, (LiveActivationError, KillSwitchError, MicroCapitalError, ReconciliationError)
    ):
        return HTTPException(
            status_code=_ERROR_STATUS.get(exc.code, 400),
            detail={"code": exc.code, "message": exc.message},
        )
    return HTTPException(status_code=500, detail="Unexpected error")


def get_activation_service(db: Session = Depends(get_db)) -> LiveActivationService:
    return LiveActivationService(db)


def get_kill_switch_service(db: Session = Depends(get_db)) -> KillSwitchService:
    return KillSwitchService(db)


def get_capital_service(db: Session = Depends(get_db)) -> MicroCapitalService:
    return MicroCapitalService(db)


def get_reconciliation_service(db: Session = Depends(get_db)) -> ReconciliationService:
    return ReconciliationService(db)


def get_micro_live_service(db: Session = Depends(get_db)) -> MicroLiveService:
    return MicroLiveService(db)


@router.get("/status", response_model=MicroLiveStatusRead)
def get_status(
    _: AuthenticatedPrincipal = Depends(RequireAnyAuthenticatedRead),
    service: MicroLiveService = Depends(get_micro_live_service),
) -> MicroLiveStatusRead:
    return MicroLiveStatusRead.model_validate(service.status().__dict__)


@router.get("/activation", response_model=ActivationStateRead)
def get_activation(
    _: AuthenticatedPrincipal = Depends(RequireAnyAuthenticatedRead),
    service: LiveActivationService = Depends(get_activation_service),
) -> ActivationStateRead:
    return ActivationStateRead.model_validate(service.status().__dict__)


@router.get("/activation/transitions", response_model=list[ActivationTransitionRead])
def list_activation_transitions(
    _: AuthenticatedPrincipal = Depends(RequireAnyAuthenticatedRead),
    service: LiveActivationService = Depends(get_activation_service),
) -> list[ActivationTransitionRead]:
    return [ActivationTransitionRead.model_validate(t) for t in service.list_transitions()]


@router.post("/activation/transition", response_model=ActivationStateRead)
def transition_activation(
    body: ActivationTransitionRequest,
    principal: AuthenticatedPrincipal = Depends(RequireFounder),
    service: LiveActivationService = Depends(get_activation_service),
) -> ActivationStateRead:
    try:
        service.transition(
            target=body.target_state,
            reason=body.reason,
            evidence=body.evidence,
            actor=principal,
        )
        return ActivationStateRead.model_validate(service.status().__dict__)
    except LiveActivationError as exc:
        raise _http(exc) from exc


@router.get("/credential-references", response_model=list[CredentialReferenceRead])
def list_credential_references(
    _: AuthenticatedPrincipal = Depends(RequireAnyAuthenticatedRead),
    service: LiveActivationService = Depends(get_activation_service),
) -> list[CredentialReferenceRead]:
    return [
        CredentialReferenceRead.model_validate(r) for r in service.list_credential_references()
    ]


@router.post(
    "/credential-references", response_model=CredentialReferenceRead, status_code=201
)
def create_credential_reference(
    body: CredentialReferenceCreate,
    principal: AuthenticatedPrincipal = Depends(RequireFounder),
    service: LiveActivationService = Depends(get_activation_service),
) -> CredentialReferenceRead:
    try:
        row = service.create_credential_reference(
            provider_key=body.provider_key,
            ref_name=body.ref_name,
            purpose=body.purpose,
            actor=principal,
        )
        return CredentialReferenceRead.model_validate(row)
    except LiveActivationError as exc:
        raise _http(exc) from exc


@router.post(
    "/credential-references/{reference_id}/validate",
    response_model=CredentialReferenceRead,
)
def validate_credential_reference(
    reference_id: uuid.UUID,
    principal: AuthenticatedPrincipal = Depends(RequireFounder),
    service: LiveActivationService = Depends(get_activation_service),
) -> CredentialReferenceRead:
    try:
        row = service.validate_credential_reference(reference_id, actor=principal)
        return CredentialReferenceRead.model_validate(row)
    except LiveActivationError as exc:
        raise _http(exc) from exc


@router.get("/kill-switches", response_model=list[KillSwitchRead])
def list_kill_switches(
    _: AuthenticatedPrincipal = Depends(RequireAnyAuthenticatedRead),
    service: KillSwitchService = Depends(get_kill_switch_service),
) -> list[KillSwitchRead]:
    return [KillSwitchRead.model_validate(s) for s in service.list_switches()]


@router.post("/kill-switches", response_model=KillSwitchRead)
def set_kill_switch(
    body: KillSwitchSet,
    principal: AuthenticatedPrincipal = Depends(RequireFounder),
    service: KillSwitchService = Depends(get_kill_switch_service),
) -> KillSwitchRead:
    try:
        row = service.set_switch(
            scope_type=body.scope_type,
            scope_id=body.scope_id,
            active=body.active,
            reason=body.reason,
            actor=principal,
        )
        return KillSwitchRead.model_validate(row)
    except KillSwitchError as exc:
        raise _http(exc) from exc


@router.get("/capital-policy", response_model=MicroCapitalPolicyRead)
def get_capital_policy(
    _: AuthenticatedPrincipal = Depends(RequireAnyAuthenticatedRead),
    service: MicroCapitalService = Depends(get_capital_service),
) -> MicroCapitalPolicyRead:
    try:
        return MicroCapitalPolicyRead.model_validate(service.get_active_policy())
    except MicroCapitalError as exc:
        raise _http(exc) from exc


@router.put("/capital-policy", response_model=MicroCapitalPolicyRead)
def put_capital_policy(
    body: MicroCapitalPolicyUpsert,
    principal: AuthenticatedPrincipal = Depends(RequireFounder),
    service: MicroCapitalService = Depends(get_capital_service),
) -> MicroCapitalPolicyRead:
    try:
        row = service.set_policy(
            max_deployable_capital=body.max_deployable_capital,
            max_order_notional=body.max_order_notional,
            max_daily_loss=body.max_daily_loss,
            max_concurrent_exposure=body.max_concurrent_exposure,
            max_provider_exposure=body.max_provider_exposure,
            max_strategy_exposure=body.max_strategy_exposure,
            actor=principal,
        )
        return MicroCapitalPolicyRead.model_validate(row)
    except MicroCapitalError as exc:
        raise _http(exc) from exc


@router.post("/reconciliation/runs", response_model=ReconciliationRunRead, status_code=201)
def create_reconciliation_run(
    body: ReconciliationRunCreate,
    principal: AuthenticatedPrincipal = Depends(RequireFounderOrOperator),
    service: ReconciliationService = Depends(get_reconciliation_service),
) -> ReconciliationRunRead:
    try:
        row = service.run(
            provider_key=body.provider_key,
            authoritative_state=body.authoritative_state,
            comparison_state=body.comparison_state,
            actor=principal,
        )
        return ReconciliationRunRead.model_validate(row)
    except ReconciliationError as exc:
        raise _http(exc) from exc


@router.get("/reconciliation/runs", response_model=list[ReconciliationRunRead])
def list_reconciliation_runs(
    _: AuthenticatedPrincipal = Depends(RequireAnyAuthenticatedRead),
    service: ReconciliationService = Depends(get_reconciliation_service),
) -> list[ReconciliationRunRead]:
    return [ReconciliationRunRead.model_validate(r) for r in service.list_runs()]


@router.get(
    "/reconciliation/runs/{run_id}/discrepancies",
    response_model=list[ReconciliationDiscrepancyRead],
)
def list_reconciliation_discrepancies(
    run_id: uuid.UUID,
    _: AuthenticatedPrincipal = Depends(RequireAnyAuthenticatedRead),
    service: ReconciliationService = Depends(get_reconciliation_service),
) -> list[ReconciliationDiscrepancyRead]:
    try:
        service.get_run(run_id)
        return [
            ReconciliationDiscrepancyRead.model_validate(d)
            for d in service.list_discrepancies(run_id)
        ]
    except ReconciliationError as exc:
        raise _http(exc) from exc


@router.get("/adapters", response_model=list[AdapterRead])
def list_adapters(
    _: AuthenticatedPrincipal = Depends(RequireAnyAuthenticatedRead),
    service: MicroLiveService = Depends(get_micro_live_service),
) -> list[AdapterRead]:
    return [AdapterRead.model_validate(p) for p, _health in service.list_adapters()]


@router.post("/dry-run/validate-order", response_model=DryRunOrderResult)
def dry_run_validate_order(
    body: DryRunOrderValidate,
    _: AuthenticatedPrincipal = Depends(RequireAnyAuthenticatedRead),
    service: MicroLiveService = Depends(get_micro_live_service),
) -> DryRunOrderResult:
    result = service.dry_run_validate_order(
        quantity=body.quantity, reference_price=body.reference_price
    )
    return DryRunOrderResult.model_validate(result)
